DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs

```diff
@@ -0,0 +1,191 @@
+                {
+                    build.Level = fortConfig.Level;
+                    build.BuildStartDate = DateTimeOffset.UnixEpoch;
+                    build.BuildEndDate = DateTimeOffset.UnixEpoch;
+                }
+            }
+        }
+    }
+
+    public async Task IncreasePlayerLevel()
+    {
+        const int MaxLevel = 60;
+        const int MaxExp = 69990;
+        DbPlayerUserData data = await userDataRepository.GetUserDataAsync();
+        data.TutorialFlag = 16640603;
+        data.TutorialStatus = 60999;
+        data.StaminaSingle = 999;
+        data.StaminaMulti = 99;
+
+        if (data.Exp < MaxExp)
+        {
+            data.Exp = MaxExp;
+        }
+
+        if (data.Level < MaxLevel)
+        {
+            data.Level = MaxLevel;
+        }
+    }
+
+    public async Task OpenTreasure(int questTreasureId, CancellationToken cancellationToken)
+    {
+        IQueryable<DbQuestTreasureList> questTreasureLists = questRepository.QuestTreasureList;
+        if (!questTreasureLists.Any(x => x.QuestTreasureId == questTreasureId))
+        {
+            QuestOpenTreasureRequest req = new() { QuestTreasureId = questTreasureId };
+            await questTreasureService.DoOpenTreasure(req, cancellationToken);
+        }
+    }
+
+    public async Task ProcessQuestCompletion(int questId)
+    {
+        DbQuest questData = await questRepository.GetQuestDataAsync(questId);
+
+        bool isFirstClear = questData.State < 3;
+        if (isFirstClear)
+            await questCompletionService.GrantFirstClearRewards(questData.QuestId);
+
+        bool[] oldMissionStatus =
+        {
+            questData.IsMissionClear1,
+            questData.IsMissionClear2,
+            questData.IsMissionClear3
+        };
+
+        QuestMissionStatus status = await CompleteQuestMissions(
+            questId,
+            oldMissionStatus
+        );
+
+        questData.IsMissionClear1 = status.Missions[0];
+        questData.IsMissionClear2 = status.Missions[1];
+        questData.IsMissionClear3 = status.Missions[2];
+
+        if (questData.BestClearTime == -1)
+        {
+            questData.BestClearTime = 36000;
+        }
+
+        questData.PlayCount += 1;
+        questData.DailyPlayCount += 1;
+        questData.WeeklyPlayCount += 1;
+        questData.State = 3;
+    }
+
+    public async Task ReadStory(int questStoryId, CancellationToken cancellationToken)
+    {
+        IEnumerable<AtgenBuildEventRewardEntityList> rewardList = await storyService.ReadStory(
+            StoryTypes.Quest,
+            questStoryId
+        );
+
+        EntityResult entityResult = StoryService.GetEntityResult(rewardList);
+        IEnumerable<AtgenQuestStoryRewardList> questRewardList = rewardList.Select(
+            StoryService.ToQuestStoryReward
+        );
+
+        UpdateDataList updateDataList = await updateDataService.SaveChangesAsync(
+            cancellationToken
+        );
```

Comment:
Why do you need this `SaveChanges`? Calling it in a loop will be very expensive

Reply:
Maybe you can help me understand database updates better. Should I only call the update service at the end after all data changes were made?